drivers/gles3/rasterizer_scene_gles3.cpp

```diff
@@ -1645,7 +1645,7 @@ void RasterizerSceneGLES3::_setup_environment(const RenderDataGLES3 *p_render_da
 		scene_state.ubo.fog_density = environment_get_fog_density(p_render_data->environment);
 		scene_state.ubo.fog_height = environment_get_fog_height(p_render_data->environment);
 		scene_state.ubo.fog_depth_curve = environment_get_fog_depth_curve(p_render_data->environment);
-		scene_state.ubo.fog_depth_begin = environment_get_fog_depth_begin(p_render_data->environment);
+		scene_state.ubo.fog_depth_begin = MIN(environment_get_fog_depth_begin(p_render_data->environment),environment_get_fog_depth_end(p_render_data->environment));
```

Comment:
This won't:
* Catch if the value is negative, and uses `far`
* Will set it to the end value, so that will not ensure this

I'd suggest:
```cpp
scene_state.ubo.fog_depth_end = environment_get_fog_depth_end(p_render_data->environment) > 0.0 ? environment_get_fog_depth_end(p_render_data->environment) : scene_state.ubo.z_far;
scene_state.ubo.fog_depth_begin = MIN(environment_get_fog_depth_begin(p_render_data->environment), scene_state.ubo.fog_depth_end - 0.001);
```
Or similar